From f4919d5acb32dd9aff35f9e4ec9e4ec4276a5c5b Mon Sep 17 00:00:00 2001
From: Jiri Olsa <jolsa@redhat.com>
Date: Mon, 8 Sep 2014 13:08:22 -0400
Subject: [tools] perf/ui/gtk: Fix callchain display

Message-id: <1410181752-25631-258-git-send-email-jolsa@redhat.com>
Patchwork-id: 91932
O-Subject: [PATCH RHEL7.1 BZ1134356 257/307] perf ui/gtk: Fix callchain display
Bugzilla: 1134356
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Don Zickus <dzickus@redhat.com>

From: Jiri Olsa <jolsa@kernel.org>

Bugzilla: 1134356
https://bugzilla.redhat.com/show_bug.cgi?id=1134356

upstream
========
commit e4cf6f886f3158061fb589df9ed452f9b30f67f1
Author: Namhyung Kim <namhyung@kernel.org>
Date: Fri May 23 18:49:33 2014 +0900

description
===========
With current output field change, GTK browser cannot display callchain
information correctly since it couldn't determine where the symbol
column is.  This is a problem - just for now I changed to use the last
column since it'll work for most cases.

Also it has a same problem of the percentage as stdio code.
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/tools/perf/ui/gtk/hists.c b/tools/perf/ui/gtk/hists.c
index 03d6812..6ca60e4 100644
--- a/tools/perf/ui/gtk/hists.c
+++ b/tools/perf/ui/gtk/hists.c
@@ -198,6 +198,13 @@ static void perf_gtk__show_hists(GtkWidget *window, struct hists *hists,
   if (perf_hpp__should_skip(fmt))
    continue;
 
+  /*
+   * XXX no way to determine where symcol column is..
+   *     Just use last column for now.
+   */
+  if (perf_hpp__is_sort_entry(fmt))
+   sym_col = col_idx;
+
   fmt->header(fmt, &hpp, hists_to_evsel(hists));
 
   gtk_tree_view_insert_column_with_attributes(GTK_TREE_VIEW(view),
@@ -253,7 +260,8 @@ static void perf_gtk__show_hists(GtkWidget *window, struct hists *hists,
 
   if (symbol_conf.use_callchain && sort__has_sym) {
    if (callchain_param.mode == CHAIN_GRAPH_REL)
-    total = h->stat.period;
+    total = symbol_conf.cumulate_callchain ?
+     h->stat_acc->period : h->stat.period;
 
    perf_gtk__add_callchain(&h->sorted_chain, store, &iter,
       sym_col, total);
-- 
1.7.1